src/Symfony/SymfonyContainerResultCacheMetaExtension.php

```diff
@@ -0,0 +1,54 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Symfony;
+
+use PHPStan\Analyser\ResultCache\ResultCacheMetaExtension;
+use function array_map;
+use function hash;
+use function serialize;
+
+final class SymfonyContainerResultCacheMetaExtension implements ResultCacheMetaExtension
+{
+
+	private ParameterMapFactory $parameterMapFactory;
+
+	private ServiceMapFactory $serviceMapFactory;
+
+	public function __construct(
+		ParameterMapFactory $parameterMapFactory,
+		ServiceMapFactory $serviceMapFactory
+	)
+	{
+		$this->parameterMapFactory = $parameterMapFactory;
+		$this->serviceMapFactory = $serviceMapFactory;
+	}
+
+	public function getKey(): string
+	{
+		return 'symfonyDiContainer';
+	}
+
+	public function getHash(): string
+	{
+		return hash('sha256', serialize([
```

Comment:
I'm not a huge fan of serialization like this. Something unexpected might occur there, some values that's not possible to serialize. What about just hashing the contents of the XML?

Reply:
As I wrote in the description (well, in the linked comment), interacting directly with XML has the drawback that result cache can be invalidated even if the container wasn't changed, only params/services were moved to other place of XML, as the final output depends on the order of source definition files. By calculating hash from parsed and sorted params/services we can be sure it will only invalidated when DI definition was _actually changed_.

Reply:
Oh I see, you're already making the scalar values here. So there's not going to be any surprise like getting the actual service objects and trying to serialize them.

I'd just exchange the `serialize` for something like `var_export` - ResultCacheManager uses that in many places.

And ask for ServiceMap/ParameterMap in the constructor instead of the factories. Thanks.

Reply:
Maps are injected directly now.